r-uefi-pi/src/fv.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright © 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use r_efi::efi::Guid;
+
+pub type FvbAttributes2 = u32;
+
+pub const FVH_SIGNATURE: u32 = 0x4856465F; // '_','F','V','H'
+use scroll::{Pread, Pwrite};
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug, Pread, Pwrite, Default)]
+pub struct FirmwareVolumeHeader {
+    pub zero_vector: [u8; 16],
+    pub file_system_guid: [u8; 16], // Guid
+    pub fv_length: u64,
+    pub signature: u32,
+    pub attributes: FvbAttributes2,
+    pub header_length: u16,
+    pub checksum: u16,
+    pub ext_header_offset: u16,
+    pub reserved: u8,
+    pub revision: u8,
+}
```

Comment:
Comments about the whole PR instead of the code piece here:)
There are many data structures defined according to some specifications, so it would be great to add methods to those data structure to verify whether its content confirms to the spec or not.

That is to enforce the rule at data structure definition instead of relying on the caller to validate the content.

Reply:
Sorry, I am not clear on what methods. Any example?
Maybe file issue to address.

Reply:
Something like this: https://github.com/jyao1/td-shim/pull/2/commits/618477058c3be8933021b3d01ab58aae73a10084#diff-a3c0fb81e06eebc8a4dad5628969726873496cbe098f40ad328e6d57e014b01eR59